tianshou/policy/modelfree/iqn.py

```diff
@@ -0,0 +1,94 @@
+import torch
+import numpy as np
+import torch.nn.functional as F
+from typing import Any, Dict
+
+from tianshou.policy import QRDQNPolicy
+from tianshou.data import Batch
+
+
+class IQNPolicy(QRDQNPolicy):
+    """Implementation of Implicit Quantile Network. arXiv:1806.06923.
+
+    :param torch.nn.Module model: a model following the rules in
+        :class:`~tianshou.policy.BasePolicy`. (s -> logits)
+    :param torch.optim.Optimizer optim: a torch.optim for optimizing the model.
+    :param float discount_factor: in [0, 1].
+    :param int sample_size: the number of samples for policy evaluation.
+        Default to 32.
+    :param int online_sample_size: the number of samples for online model
+        in training. Default to 8.
+    :param int target_sample_size: the number of samples for target model
+        in training. Default to 8.
+    :param int estimation_step: the number of steps to look ahead. Default to 1.
+    :param int target_update_freq: the target network update frequency (0 if
+        you do not use the target network).
+    :param bool reward_normalization: normalize the reward to Normal(0, 1).
+        Default to False.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.QRDQNPolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        discount_factor: float = 0.99,
+        sample_size: int = 32,
+        online_sample_size: int = 8,
+        target_sample_size: int = 8,
+        estimation_step: int = 1,
+        target_update_freq: int = 0,
+        reward_normalization: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(model, optim, discount_factor, sample_size, estimation_step,
+                         target_update_freq, reward_normalization, **kwargs)
+        assert sample_size > 1, "sample_size should be greater than 1"
+        assert online_sample_size > 1, "online_sample_size should be greater than 1"
+        assert target_sample_size > 1, "target_sample_size should be greater than 1"
+        self._sample_size = sample_size  # for policy eval
+        self._online_sample_size = online_sample_size
+        self._target_sample_size = target_sample_size
+        # set sample size for online and target model
+        self.model.sample_size = self._online_sample_size  # type: ignore
```

Comment:
plz do not do this magic
is it possible to add an extra argument in net.forward?

Reply:
Good point. I did consider adding it as an extra argument in `forward()` but somehow thought it was more complicated. Now I look at it again and it seems I should put `sample_size` completely in the model, instead of in the policy. I'll make the change.